README.md

```diff
@@ -0,0 +1,207 @@
+
+### Library
+
+The library is the core of the server, which does the following:
+ - implements process management APIs:
+   - `StartProcess` starts a new process.
+   - `GetProcessStatus` returns process status.
+   - `StreamOutput` returns combined process standard and error output stream.
+   - `StopProcess` stops the process.
+ - verifies user authorization:
+   - the library maintains an authorization table of clients and corresponding bitmap of permitted APIs.
+   - at the beginning of each API call the library checks that the client is authorized to call the API.
+   - for `GetProcessStatus`, `StreamOutput`, and `StopProcess` APIs, the library verifies that the process has been created by the same client.
+   - *note:* the library does not have the notion of the `superuser`, but it can be easily added.
+ - implements resource control for the processes.
+ - generates UUID for processes (`github.com/google/uuid`).
+ - maintains a process table, for both active and terminated processes. The key is process UUID. The values is a structure representing the process:
+ ```
+ type Process struct {
+	clientID string        // client ID retrieved from the client certificate
+	cmd      *exec.Cmd     // process object in Go
+	output   bytes.Buffer  // buffer containing combined process standard and error outputs
+	status   *proto.Status // process status
+}
+ ```
+
+The library exposes singleton process manager object holding all aforementioned structures:
+```
+const (
+	PermStart  = 0x01
+	PermStop   = 0x02
+	PermStatus = 0x04
+	PermStream = 0x08
+)
+
+type ProcManager struct {
+	// process table [process UUID : Process]
+	procs     map[string]*Process
+	procMutex sync.Mutex
+
+	// permission table [client ID : permission bitmap]
+	perm map[string]int
+
+	// process UUID
+	uuid string
+}
+```
+#### Resource control
+
+The control over compute, disk I/O and memory resources is provided by means of cgroups.
+When the library receives an API call to start a process, it will create the following files:
+- `/sys/fs/cgroup/cpu/worker-<UUID>/cpu.shares`. This file contains value that limit number of CPU share for the process.
+- `/sys/fs/cgroup/memory/worker-<UUID>/memory.limit_in_bytes`. This file contains value that limits the amount of the virtual and physical memory for the process.
+- `/sys/fs/cgroup/blkio/worker-<UUID>/blkio.throttle.read_bps_device`. This file contains list of block devices followed by a value that limits the read bandwidth rate for the process.
+- `/sys/fs/cgroup/blkio/worker-<UUID>/blkio.throttle.write_bps_device`. This file contains list of block devices followed by a value that limits the write bandwidth rate for the process.
+
+Once the process is successfully started, the library will create the following files, containing PID of that process (`process.cmd.Process.Pid`)
+- `/sys/fs/cgroup/cpu/worker-<UUID>/cgroup.procs`
+- `/sys/fs/cgroup/memory/worker-<UUID>/cgroup.procs`
+- `/sys/fs/cgroup/blkio/worker-<UUID>/cgroup.procs`
```

Comment:
Based on current description a process run for short period with unlimited resource, also if a process forks before a process PID is written to cgroups it will escape resource control.

Reply:
Oh, I see. Then I would need to start an external program that creates cgroup directories, writes its PID in `cgroup.procs` files, then clones itself with the same PID and runs the command. Is that the general direction?

Reply:
@smallinsky I ran a small experiment and wrote PID after the process was forked.
It seems to work with regards to the resource control: https://github.com/dmitsh/cgrouptest

Reply:
>  I ran a small experiment and wrote PID after the process was forked.
It seems to work with regards to the resource control: https://github.com/dmitsh/cgrouptest
Both `runner` and `runner2` example writes the parent PID before to cgroups before child process is started: 
https://github.com/dmitsh/cgrouptest/blob/main/cmd/runner/main.go#L30

if you modify the example and move Parent PID write to cgorups after cmd.Start call and remove child process child PID cgroups write call  the child process will run with unlimited memory. 


Could you address this issue in the document ?

Reply:
@smallinsky I'm not following. Could you elaborate?

Reply:
@dmitsh 
The second solution presented in runner2 dir solves the cgroups escape problem and limits the process resource from the beginning. Could you updated the design document based on `cgrouptest/runner2` implementation ?

Regarding [runner](https://github.com/dmitsh/cgrouptest/blob/main/cmd/runner2/main.go) solution, I think that it works because child process "app" doesn't fork.

Reply:
I have already updated the README per `runner2` example.

Reply:
Ah, Sorry I have missed this doc updated  🙃

Reply:
Thanks! I'm resolving this comment then.